agent/api/appmesh/appmesh.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package appmesh
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+const (
+	appMesh            = "APPMESH"
+	splitter           = ","
+	ignoredUID         = "IgnoredUID"
+	ignoredGID         = "IgnoredGID"
+	proxyIngressPort   = "ProxyIngressPort"
+	proxyEgressPort    = "ProxyEgressPort"
+	appPorts           = "AppPorts"
+	egressIgnoredIPs   = "EgressIgnoredIPs"
+	egressIgnoredPorts = "EgressIgnoredPorts"
+)
+
+// AppMesh contains information of app mesh config
+type AppMesh struct {
+	// IgnoredUID is egress traffic from the processes owned by the UID will be ignored
+	IgnoredUID string
+	// IgnoredGID specifies egress traffic from the processes owned by the GID will be ignored
+	IgnoredGID string
+	// ProxyIngressPort is the ingress port number that proxy is listening on
+	ProxyIngressPort string
+	// ProxyEgressPort is the egress port number that proxy is listening on
+	ProxyEgressPort string
+	// AppPorts is the port number that application is listening on
+	AppPorts []string
+	// EgressIgnoredIPs is the list of ports for which egress traffic will be ignored
+	EgressIgnoredIPs []string
+	// EgressIgnoredPorts is the list of IPs for which egress traffic will be ignored
+	EgressIgnoredPorts []string
+}
+
+// AppMeshFromACS validates proxy config if it is app mesh type and create AppMesh object
+func AppMeshFromACS(proxyConfig *ecsacs.ProxyConfiguration) (*AppMesh, error) {
+
+	if proxyConfig == nil {
+		return nil, nil
+	}
+
+	if *proxyConfig.Type != appMesh {
+		return nil, fmt.Errorf("agent does not support proxy type other than app mesh")
```

Comment:
The function name is AppMeshFromACS, should we check the proxyConfig.Type outside this function? Like only when proxyConfig isn't nil and Type == appMesh, we call this AppMeshFromACS. Otherwise, we return the error. I don't feel it is the right place to have the validation of type here.

Reply:
If you decide to move the validation outside the function, remember to resolve aaithal@ comment and remove error in return

Reply:
I agree if we have multiple proxy support we should not validate here(Prob not in payload handler, we should prob have some sort of proxy constructor file). But I intentionally put the appmesh type check here to make it clear in agent proxy configuration is strongly typed with appmesh struct . I am thinking if in foreseeable future we won't have any other proxy, appmesh struct is strongly typed with proxy configuration struct. we should make it clear in the code and in a centralized place when we construct appmesh struct, we should throw exceptions instead of same logic for appmesh spanned in multiple files. How do you think?